Add doc for torch.cond

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108691

We add a doc for torch.cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977.

ydwu4 committed:
Add doc for torch.ops.higher_order.cond

[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.ops.higher_order.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

zou3519 commented on docs/source/control_flow_cond.rst:
```diff
@@ -0,0 +1,150 @@
+
+Export supports the structured control flow operator `torch.ops.higher_order.cond`. It can be used to specify if-else like control flow within
+some code.
+
+torch.ops.higher_order.cond
```
Do we really need to document this? Ideally we would replace this with torch.cond, but that didn't make the branch cut. In the future we want people to use torch.cond instead of directly using torch.ops.higher_order.cond.

If we must document this, then there needs to be a big note somewhere that this API is "prototype" as a part of the prototype torch.export release.

ydwu4 replied:
Yeah, these are valid points. I also want to add that people will be confused about which api to use. Ideally, there should be one and only one way to do conditional, that is, torch.cond. When torch.cond is ready, we can have a documentation for the API torch.cond and an IR node specification for torch.ops.higher_order.cond so that users can do further transformations/lowering.  For now, it might do more harm if we release a documentation that directly uses torch.ops.higher_order.cond . 

What do you think? @gmagogsfm

gmagogsfm replied:
> Do we really need to document this?

It depends on whether there are flaws with the part of cond implementation that made branch cut. Will there be adverse effects (other than API confusion) if we make `torch.ops.higher_order.cond` public? 

> If we must document this, then there needs to be a big note somewhere that this API is "prototype" as a part of the prototype torch.export release.
Agreed

ydwu4 committed:
Update on "Add doc for torch.ops.higher_order.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.ops.higher_order.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

zou3519 reviewed (approved):
The warning is good enough for me

angelayi commented on docs/source/export.rst:
```diff
@@ -31,3 +31,4 @@ torch.export
    :maxdepth: 1
 
    generated/exportdb/index
+   control_flow_cond
```
Could you add the link [here](https://github.com/pytorch/pytorch/blob/bde75eb9a8035e369fc0964c2cc604c7b0df8349/docs/source/export.rst?plain=1#L500) and [here](https://github.com/pytorch/pytorch/blob/bde75eb9a8035e369fc0964c2cc604c7b0df8349/docs/source/export.rst?plain=1#L536)

ydwu4 committed:
Update on "Add doc for torch.ops.higher_order.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.ops.higher_order.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.ops.higher_order_cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 


[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -44,43 +44,35 @@ class UnsupportedAliasMutationException(RuntimeError):
 def cond(pred, true_fn, false_fn, operands):
     r"""
-    Conditionally applies ``true_fn`` or ``false_fn``.
+    Conditionally applies `true_fn` or `false_fn`.
```
Please keep the prototype warning. cond is not going out of prototype until we do a feature review for it (or for export) to promote to beta

ydwu4 replied:
added back.

zou3519 commented on docs/source/control_flow_cond.rst:
```diff
@@ -0,0 +1,169 @@
+operator (`torch.ops.higher_order.cond`), which preserves predicate, true function and false functions.
+This unlocks great flexibilty in writing and deploying models that change model architecture based on
+the **value** or **shape** of inputs or intermediates.
+
```
Please add a prototype warning here.

ydwu4 replied:
added.

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 



[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 



[ghstack-poisoned]

ydwu4 committed:
Update on "Add doc for torch.cond"


We add a doc for torch.cond. This PR is a replacement of https://github.com/pytorch/pytorch/pull/107977. 



[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
